src/shared/image_utils.js

```diff
@@ -13,30 +13,42 @@
  * limitations under the License.
  */
 
+import { FeatureTest } from "./util.js";
+
 function applyMaskImageData({
   src,
   srcPos = 0,
   dest,
-  destPos = 3,
+  destPos = 0,
   width,
   height,
   inverseDecode = false,
 }) {
-  const srcLength = src.byteLength;
-  const zeroMapping = inverseDecode ? 0 : 255;
-  const oneMapping = inverseDecode ? 255 : 0;
+  const opaque = FeatureTest.isLittleEndian ? 0xff000000 : 0x000000ff;
+  const [zeroMapping, oneMapping] = !inverseDecode ? [opaque, 0] : [0, opaque];
+  const widthInSource = width >> 3;
+  const widthRemainder = width & 7;
+  dest = new Uint32Array(dest.buffer);
 
-  for (let j = 0; j < height; j++) {
-    let elem,
-      mask = 0;
-    for (let k = 0; k < width; k++) {
-      if (mask === 0) {
-        elem = srcPos < srcLength ? src[srcPos++] : 255;
-        mask = 128;
-      }
-      dest[destPos] = elem & mask ? oneMapping : zeroMapping;
-      destPos += 4;
-      mask >>= 1;
+  for (let i = 0; i < height; i++) {
+    for (const max = srcPos + widthInSource; srcPos < max; srcPos++) {
+      const elem = srcPos < src.length ? src[srcPos] : 255;
+      dest[destPos] = elem & 0b10000000 ? oneMapping : zeroMapping;
+      dest[destPos + 1] = elem & 0b1000000 ? oneMapping : zeroMapping;
+      dest[destPos + 2] = elem & 0b100000 ? oneMapping : zeroMapping;
+      dest[destPos + 3] = elem & 0b10000 ? oneMapping : zeroMapping;
+      dest[destPos + 4] = elem & 0b1000 ? oneMapping : zeroMapping;
+      dest[destPos + 5] = elem & 0b100 ? oneMapping : zeroMapping;
+      dest[destPos + 6] = elem & 0b10 ? oneMapping : zeroMapping;
+      dest[destPos + 7] = elem & 0b1 ? oneMapping : zeroMapping;
+      destPos += 8;
```

Comment:
Nit: Similar to [this code in `canvas.js`](https://github.com/mozilla/pdf.js/blob/08e1abe68e5df09c753665739e09c4529e1f6e61/src/display/canvas.js#L726-L733), can we use the `dest[destPos++] = ...` format here as well since that (visually) aligns the code a tiny bit nicer?

Reply:
My goal is mainly to have some good perfs: a `dest++` induces that the new value is stored in `dest` which is useless.
With this code we've only one store.

Reply:
I just tried the dest++ and it doesn't make any significative difference, let's go for it.

Reply:
I guess the counter can probably live in a register while it's in the loop scope and then the new value is stored only at the end of the main loop.